In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
from sklearn.preprocessing import MinMaxScaler,StandardScaler,LabelEncoder,OneHotEncoder

In [7]:
# reading file
df=pd.read_csv('/content/adult_with_headers (1).csv')
df.head()

age          workclass  fnlwgt  ... hours_per_week  native_country  income
0   39          State-gov   77516  ...             40   United-States   <=50K
1   50   Self-emp-not-inc   83311  ...             13   United-States   <=50K
2   38            Private  215646  ...             40   United-States   <=50K
3   53            Private  234721  ...             40   United-States   <=50K
4   28            Private  338409  ...             40            Cuba   <=50K

[5 rows x 15 columns]

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education_num   32561 non-null  int64 
 5   marital_status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital_gain    32561 non-null  int64 
 11  capital_loss    32561 non-null  int64 
 12  hours_per_week  32561 non-null  int64 
 13  native_country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [12]:
df.describe()

age        fnlwgt  ...  capital_loss  hours_per_week
count  32561.000000  3.256100e+04  ...  32561.000000    32561.000000
mean      38.581647  1.897784e+05  ...     87.303830       40.437456
std       13.640433  1.055500e+05  ...    402.960219       12.347429
min       17.000000  1.228500e+04  ...      0.000000        1.000000
25%       28.000000  1.178270e+05  ...      0.000000       40.000000
50%       37.000000  1.783560e+05  ...      0.000000       40.000000
75%       48.000000  2.370510e+05  ...      0.000000       45.000000
max       90.000000  1.484705e+06  ...   4356.000000       99.000000

[8 rows x 6 columns]

In [13]:
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education_num     0
marital_status    0
occupation        0
relationship      0
race              0
sex               0
capital_gain      0
capital_loss      0
hours_per_week    0
native_country    0
income            0
dtype: int64

## Identify Feature Types

Separate the dataset's columns into numerical and categorical types to prepare for specific preprocessing steps.


In [29]:
numerical_cols = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_cols = df.select_dtypes(include=['object']).columns.tolist()

print("Numerical Columns:", numerical_cols)
print("Categorical Columns:", categorical_cols)

Numerical Columns: ['age', 'workclass', 'fnlwgt', 'education', 'education_num', 'marital_status', 'occupation', 'relationship', 'capital_gain', 'capital_loss', 'hours_per_week', 'native_country', 'Capital Difference', 'capital_gain_log', 'capital_loss_log']
Categorical Columns: []


## Handle Missing Values

Review the dataset for any missing values.there are currently no missing values to handle. If any were present, appropriate strategies like imputation or removal would be applied.


## Apply Standard Scaling

Apply StandardScaler to the numerical features of the dataset. This technique scales features to have a mean of 0 and a standard deviation of 1.



First, I will instantiate StandardScaler to prepare for scaling the numerical data. Then, I will apply the scaler to the numerical columns of the DataFrame df using fit_transform and update the original numerical columns with the scaled data.



In [21]:
scaler = StandardScaler()
df[numerical_cols] = scaler.fit_transform(df[numerical_cols])
print("Numerical columns after StandardScaler application:")
print(df[numerical_cols].head())

Numerical columns after StandardScaler application:
        age    fnlwgt  ...  capital_loss  hours_per_week
0  0.030671 -1.063611  ...      -0.21666       -0.035429
1  0.837109 -1.008707  ...      -0.21666       -2.222153
2 -0.042642  0.245079  ...      -0.21666       -0.035429
3  1.057047  0.425801  ...      -0.21666       -0.035429
4 -0.775768  1.408176  ...      -0.21666       -0.035429

[5 rows x 6 columns]


## Apply Min-Max Scaling

Apply MinMaxScaler to the numerical features of the dataset. This technique scales features to a fixed range, usually 0 to 1. After applying, discuss the scenarios where StandardScaler is preferred over MinMaxScaler and vice-versa.

## Apply Min-Max Scaling

Apply MinMaxScaler to the numerical features of the dataset. This technique scales features to a fixed range, usually 0 to 1. After applying, discuss the scenarios where StandardScaler is preferred over MinMaxScaler and vice-versa.


I will instantiate MinMaxScaler to prepare for scaling the numerical data. Then, I will apply the scaler to the numerical columns of the DataFrame df using fit_transform and update the original numerical columns with the scaled data.



In [22]:
scaler_minmax = MinMaxScaler()
df[numerical_cols] = scaler_minmax.fit_transform(df[numerical_cols])
print("Numerical columns after MinMaxScaler application:")
print(df[numerical_cols].head())

Numerical columns after MinMaxScaler application:
        age    fnlwgt  ...  capital_loss  hours_per_week
0  0.301370  0.044302  ...           0.0        0.397959
1  0.452055  0.048238  ...           0.0        0.122449
2  0.287671  0.138113  ...           0.0        0.397959
3  0.493151  0.151068  ...           0.0        0.397959
4  0.150685  0.221488  ...           0.0        0.397959

[5 rows x 6 columns]


### Discussion on StandardScaler vs. MinMaxScaler

**StandardScaler**
*   **Preferred when:**
    *   The data follows a Gaussian (normal) distribution or is approximately Gaussian. Many machine learning algorithms (e.g., Logistic Regression, Linear Regression, SVMs, neural networks) assume a Gaussian distribution or perform better when features are scaled to have a mean of 0 and a standard deviation of 1.
    *   Outliers are present in the data. StandardScaler is less affected by outliers because it doesn't bound the data to a specific range. While outliers will still influence the mean and standard deviation, they won't compress the majority of the data into a tiny range as much as MinMaxScaler might.
    *   The algorithm relies on distance calculations (e.g., K-Means, K-Nearest Neighbors) or gradient descent. Scaling to a standard normal distribution helps ensure that all features contribute equally to the distance metric or the gradient updates.

**MinMaxScaler**
*   **Preferred when:**
    *   The data does not follow a Gaussian distribution, and the algorithm is sensitive to the scale of the features but not necessarily their distribution (e.g., Neural Networks, algorithms that do not assume normality).
    *   You need the transformed features to be within a specific range (e.g., 0-1 or -1 to 1). This can be useful for algorithms that require input features to be positive or within a certain bounded range, such as image processing or some neural network activation functions.
    *   You want to preserve the relative relationships between data points more directly. Min-Max scaling preserves the exact relationships between data points by simply translating and scaling them.
    *   Outliers are either not present or have been handled. MinMaxScaler is very sensitive to outliers; a single outlier can shift the entire range significantly, compressing the bulk of the data into a small interval.

**Key Considerations:**
*   **Outliers:** StandardScaler is generally more robust to outliers than MinMaxScaler. If your dataset has significant outliers and you choose MinMaxScaler, you might need to handle them first (e.g., cap or remove them).
*   **Algorithm Requirements:** The choice often depends on the specific machine learning algorithm you plan to use and its assumptions about feature scaling and distribution.
*   **Interpretability:** Min-Max scaling can sometimes be more intuitive as it maps data to a clear, fixed range, making it easier to understand the scaled values relative to the maximum and minimum.

## Apply One-Hot Encoding for Categorical Features (<= 5 unique categories)

Apply One-Hot Encoding to categorical features that have 5 or fewer unique categories. This encoding method creates new binary columns for each category, which is suitable for nominal categorical data where there is no inherent order.



In [23]:
categorical_cols_ohe = [col for col in categorical_cols if df[col].nunique() <= 5]
categorical_cols_label = [col for col in categorical_cols if df[col].nunique() > 5]

print("Categorical columns for One-Hot Encoding:", categorical_cols_ohe)
print("Categorical columns for Label Encoding (more than 5 unique categories):", categorical_cols_label)

df_encoded = pd.get_dummies(df, columns=categorical_cols_ohe, drop_first=True)

print("DataFrame head after One-Hot Encoding:")
print(df_encoded.head())

df = df_encoded

Categorical columns for One-Hot Encoding: ['race', 'sex', 'income']
Categorical columns for Label Encoding (more than 5 unique categories): ['workclass', 'education', 'marital_status', 'occupation', 'relationship', 'native_country']
DataFrame head after One-Hot Encoding:
        age          workclass    fnlwgt  ... race_ White  sex_ Male income_ >50K
0  0.301370          State-gov  0.044302  ...        True       True        False
1  0.452055   Self-emp-not-inc  0.048238  ...        True       True        False
2  0.287671            Private  0.138113  ...        True       True        False
3  0.493151            Private  0.151068  ...       False       True        False
4  0.150685            Private  0.221488  ...       False      False        False

[5 rows x 18 columns]


## Apply Label Encoding for Categorical Features (> 5 unique categories)

Apply Label Encoding to categorical features that have more than 5 unique categories. This encoding method assigns a unique integer to each category, which is suitable for ordinal categorical data or when the number of unique categories is large. After applying, discuss the advantages and disadvantages of One-Hot Encoding and Label Encoding.

In [24]:
for col in categorical_cols_label:
    le = LabelEncoder()
    df[col] = le.fit_transform(df[col])
    print(f"Column '{col}' after Label Encoding:")
    print(df[col].head())

print("\nDataFrame head after Label Encoding application for relevant columns:")
print(df.head())

Column 'workclass' after Label Encoding:
0    7
1    6
2    4
3    4
4    4
Name: workclass, dtype: int64
Column 'education' after Label Encoding:
0     9
1     9
2    11
3     1
4     9
Name: education, dtype: int64
Column 'marital_status' after Label Encoding:
0    4
1    2
2    0
3    2
4    2
Name: marital_status, dtype: int64
Column 'occupation' after Label Encoding:
0     1
1     4
2     6
3     6
4    10
Name: occupation, dtype: int64
Column 'relationship' after Label Encoding:
0    1
1    0
2    1
3    0
4    5
Name: relationship, dtype: int64
Column 'native_country' after Label Encoding:
0    39
1    39
2    39
3    39
4     5
Name: native_country, dtype: int64

DataFrame head after Label Encoding application for relevant columns:
        age  workclass    fnlwgt  ...  race_ White  sex_ Male  income_ >50K
0  0.301370          7  0.044302  ...         True       True         False
1  0.452055          6  0.048238  ...         True       True         False
2  0.287671          4

### Discussion on One-Hot Encoding vs. Label Encoding

**One-Hot Encoding (OHE)**
*   **Advantages:**
    *   **No Implied Ordinality:** Creates binary columns for each category, ensuring that no artificial ordinal relationship is imposed on nominal categorical features. This is crucial for algorithms that interpret numerical values as having an ordered relationship (e.g., linear models, tree-based models).
    *   **Increased Model Performance:** Can often lead to better performance in models that expect numerical input and struggle with arbitrary integer assignments for nominal data.
    *   **Interpretability:** Each category has its own column, making it easier to interpret the impact of individual categories on the target variable.
*   **Disadvantages:**
    *   **Curse of Dimensionality:** Can significantly increase the number of features, especially if a categorical variable has many unique categories. This can lead to increased computational cost and memory usage, and potentially reduce model performance (e.g., in some distance-based algorithms).
    *   **Multicollinearity:** If not handled (e.g., by dropping one of the dummy variables), it can introduce multicollinearity, which can be an issue for some regression models.

**Label Encoding**
*   **Advantages:**
    *   **Simplicity and Compactness:** Transforms categorical data into numerical data without increasing the feature space. This is memory-efficient and computationally less expensive, especially for datasets with a high number of unique categories.
    *   **Suitable for Ordinal Data:** Ideal for categorical features that inherently have an order (e.g., "low", "medium", "high"). The numerical assignment reflects this order.
*   **Disadvantages:**
    *   **Implied Ordinality:** Imposes an arbitrary ordinal relationship on nominal categorical data. For example, if categories A, B, C are encoded as 0, 1, 2, a model might interpret C > B > A, even if no such relationship exists. This can lead to incorrect assumptions and negatively impact model performance, especially with algorithms like linear regression or SVMs.
    *   **Bias Towards Higher Values:** Models might give more weight or importance to categories with higher numerical labels, which can be misleading if the labels are arbitrary.

**When to choose which:**
*   Use **One-Hot Encoding** for nominal categorical features (no inherent order) and when the number of unique categories is relatively small.
*   Use **Label Encoding** for ordinal categorical features (with inherent order) or when the number of unique categories is very large, and the model used (e.g., tree-based models like Random Forest, XGBoost) can handle the artificial ordinality without significant impact or can inherently manage high-cardinality features.

## Create 'Age Group' Feature


Create a new categorical feature 'Age Group' by binning the 'age' column. This will help in grouping individuals into different age categories, which can be useful for analysis or modeling.

In [25]:
bins = [0, 0.2, 0.4, 0.6, 0.8, 1.0]
labels = ['Young Adult', 'Adult', 'Middle Aged', 'Senior', 'Elderly']
df['Age Group'] = pd.cut(df['age'], bins=bins, labels=labels, right=False)

print("DataFrame head with new 'Age Group' feature:")
print(df[['age', 'Age Group']].head())
print("\nValue counts for 'Age Group':")
print(df['Age Group'].value_counts())

DataFrame head with new 'Age Group' feature:
        age    Age Group
0  0.301370        Adult
1  0.452055  Middle Aged
2  0.287671        Adult
3  0.493151  Middle Aged
4  0.150685  Young Adult

Value counts for 'Age Group':
Age Group
Adult          12211
Young Adult    11460
Middle Aged     6558
Senior          2091
Elderly          198
Name: count, dtype: int64


## Create New Feature: Capital Difference

Derive a new numerical feature by calculating the difference between 'capital_gain' and 'capital_loss' to represent the net capital activity.


In [26]:
df['Capital Difference'] = df['capital_gain'] - df['capital_loss']

print("DataFrame head with new 'Capital Difference' feature:")
print(df[['capital_gain', 'capital_loss', 'Capital Difference']].head())

DataFrame head with new 'Capital Difference' feature:
   capital_gain  capital_loss  Capital Difference
0       0.02174           0.0             0.02174
1       0.00000           0.0             0.00000
2       0.00000           0.0             0.00000
3       0.00000           0.0             0.00000
4       0.00000           0.0             0.00000


In [27]:
import numpy as np

# Identify skewed numerical features. 'capital_gain' and 'capital_loss' are typically highly skewed with many zeros.
# Since the numerical columns were already MinMaxScaled to [0,1], applying log1p will transform values from [0,1] to [0, log(2)].
# This still helps in reducing skewness within the transformed range.
df['capital_gain_log'] = np.log1p(df['capital_gain'])
df['capital_loss_log'] = np.log1p(df['capital_loss'])

print("Original 'capital_gain' and 'capital_loss' (MinMax scaled) and their log-transformed versions:")
print(df[['capital_gain', 'capital_gain_log', 'capital_loss', 'capital_loss_log']].head())

Original 'capital_gain' and 'capital_loss' (MinMax scaled) and their log-transformed versions:
   capital_gain  capital_gain_log  capital_loss  capital_loss_log
0       0.02174          0.021507           0.0               0.0
1       0.00000          0.000000           0.0               0.0
2       0.00000          0.000000           0.0               0.0
3       0.00000          0.000000           0.0               0.0
4       0.00000          0.000000           0.0               0.0


## Summary:

### Data Analysis Key Findings

*   **Feature Identification**: The dataset `df` contains 6 numerical columns (`age`, `fnlwgt`, `education_num`, `capital_gain`, `capital_loss`, `hours_per_week`) and 9 categorical columns (`workclass`, `education`, `marital_status`, `occupation`, `relationship`, `race`, `sex`, `native_country`, `income`).
*   **Missing Values**: No missing values were found in the dataset, eliminating the need for imputation or removal.
*   **Numerical Feature Scaling**:
    *   All numerical features were initially scaled using `StandardScaler` (mean 0, standard deviation 1).
    *   Subsequently, `MinMaxScaler` was applied to the same numerical features, scaling them to a range between 0 and 1. The Min-Max scaled values effectively replaced the Standard Scaled values.
*   **Categorical Feature Encoding**:
    *   **One-Hot Encoding**: Categorical features with 5 or fewer unique categories (`race`, `sex`, `income`) were One-Hot Encoded, creating binary columns for each category.
    *   **Label Encoding**: Categorical features with more than 5 unique categories (`workclass`, `education`, `marital_status`, `occupation`, `relationship`, `native_country`) were Label Encoded, assigning a unique integer to each category.
*   **Feature Engineering - 'Age Group'**: A new categorical feature 'Age Group' was created by binning the scaled 'age' column into five categories: 'Young Adult', 'Adult', 'Middle Aged', 'Senior', and 'Elderly'. The distribution of these groups was: Adult (12,211), Young Adult (11,460), Middle Aged (6,558), Senior (2,091), and Elderly (198).
*   **Feature Engineering - 'Capital Difference'**: A new numerical feature 'Capital Difference' was derived by subtracting 'capital\_loss' from 'capital\_gain', representing net capital activity.
*   **Skewness Handling**: The `capital_gain` and `capital_loss` features were identified as potentially skewed and were log-transformed using `np.log1p` to reduce skewness, creating `capital_gain_log` and `capital_loss_log`.

### Insights or Next Steps

*   The choice between `StandardScaler` and `MinMaxScaler` or One-Hot Encoding and Label Encoding depends heavily on the specific machine learning algorithms to be used and the nature of the data (e.g., presence of outliers, data distribution, cardinality). This comprehensive preprocessing prepares the dataset for a wide array of modeling approaches.
*   Further analysis could involve exploring the distributions of the newly created and transformed features, particularly 'Age Group' and 'Capital Difference', to understand their relationship with the target variable (`income`). This could lead to further feature selection or weighting strategies before model training.
